### ISSUE TITLE
Each time a pod exits, kubelet cpu usage increases with cgroups v2

### ISSUE THREAD

**Issue by MatthewFagan** [kubernetes/kubernetes#112151]
### What happened?

When a pod exits, there is some problem in the cgroup cleanup. The containerd cgroup within the kubelet cgroup does not get deleted, and then the kubelet keeps trying to delete the kubelet cgroup, which fails because of the child cgroup. The kubelet continues to try to delete the cgroup every 1 second or so. As pods start and stop, the number of cgroups in this state grows, and the kubelet process itself gradually consumes more and more cpu, until it gets to 100% of available.

Here are snippets of the /var/log/syslog for a single example. This is for the case of kubelet cgroup /sys/fs/cgroup/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podde737447_24b2_4f9e_a805_ad1161a169fb.slice which contains a sub-cgroup cri-containerd-67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7.scope.

First, logs of the containerd startup/shutdown:

```console
Aug 31 04:32:02 ip-10-128-105-64 containerd[4512]: time="2022-08-31T04:32:02.782509663Z" level=info msg="CreateContainer within sandbox \"5e10d7b5e59de1b275ac54c1e1b60c33d50388246339e7390700fd44bc0a3946\" for &ContainerMetadata{Name:workspace,Attempt:0,} retu
rns container id \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\""
Aug 31 04:32:02 ip-10-128-105-64 containerd[4512]: time="2022-08-31T04:32:02.782994283Z" level=info msg="StartContainer for \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\""
Aug 31 04:32:02 ip-10-128-105-64 systemd[1]: Started libcontainer container 67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7.
Aug 31 04:32:02 ip-10-128-105-64 containerd[4512]: time="2022-08-31T04:32:02.932528287Z" level=info msg="StartContainer for \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\" returns successfully"
Aug 31 04:32:03 ip-10-128-105-64 kubelet[4540]: I0831 04:32:03.878016    4540 kubelet.go:2145] "SyncLoop (PLEG): event for pod" pod="default/ws-6d246aa1-7b06-4d20-bb4d-b97a5ee23ce3" event=&{ID:de737447-24b2-4f9e-a805-ad1161a169fb Type:ContainerStarted Data:67
620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7}
Aug 31 04:32:41 ip-10-128-105-64 containerd[4512]: time="2022-08-31T04:32:41.044475821Z" level=error msg="Failed to get usage for snapshot \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\"" error="lstat /var/lib/containerd/io.containerd.sna
pshotter.v1.overlayfs/snapshots/302/fs/home/gitpod/.pnpm-store/v3/files/4e/_tmp_2354_b85efb9d908a56a70664bfc205d75ad1: no such file or directory"
Aug 31 04:32:51 ip-10-128-105-64 containerd[4512]: time="2022-08-31T04:32:51.121230307Z" level=error msg="Failed to get usage for snapshot \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\"" error="lstat /var/lib/containerd/io.containerd.sna
pshotter.v1.overlayfs/snapshots/302/fs/home/gitpod/.pnpm-store/v3/files/6d/_tmp_2354_586dd8b5a753516c4f6b1e08ce0eb8d9: no such file or directory"
Aug 31 05:02:30 ip-10-128-105-64 kubelet[4540]: I0831 05:02:30.445889    4540 kuberuntime_container.go:723] "Killing container with a grace period" pod="default/ws-6d246aa1-7b06-4d20-bb4d-b97a5ee23ce3" podUID=de737447-24b2-4f9e-a805-ad1161a169fb containerName
="workspace" containerID="containerd://67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7" gracePeriod=30
Aug 31 05:02:30 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:30.446103871Z" level=info msg="StopContainer for \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\" with timeout 30 (s)"
Aug 31 05:02:30 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:30.446457870Z" level=info msg="Stop container \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\" with signal terminated"
Aug 31 05:02:31 ip-10-128-105-64 systemd[1]: cri-containerd-67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7.scope: Succeeded.
Aug 31 05:02:31 ip-10-128-105-64 systemd[1]: cri-containerd-67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7.scope: Consumed 20min 19.229s CPU time.
Aug 31 05:02:31 ip-10-128-105-64 systemd[47059]: run-containerd-io.containerd.runtime.v2.task-k8s.io-67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7-rootfs.mount: Succeeded.
Aug 31 05:02:31 ip-10-128-105-64 systemd[1]: run-containerd-io.containerd.runtime.v2.task-k8s.io-67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7-rootfs.mount: Succeeded.
Aug 31 05:02:31 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:31.479369518Z" level=info msg="shim disconnected" id=67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7
Aug 31 05:02:31 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:31.479809566Z" level=warning msg="cleaning up after shim disconnected" id=67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7 namespace=k8s.io
Aug 31 05:02:31 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:31.509386019Z" level=info msg="StopContainer for \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\" returns successfully"
Aug 31 05:02:31 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:31.509958931Z" level=info msg="Container to stop \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\" must be in running or unknown state, current state \"CONTAINER_EXIT
ED\""
Aug 31 05:02:32 ip-10-128-105-64 kubelet[4540]: I0831 05:02:32.453156    4540 generic.go:296] "Generic (PLEG): container finished" podID=de737447-24b2-4f9e-a805-ad1161a169fb containerID="67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7" exitCo
de=0
Aug 31 05:02:32 ip-10-128-105-64 kubelet[4540]: I0831 05:02:32.453199    4540 kubelet.go:2145] "SyncLoop (PLEG): event for pod" pod="default/ws-6d246aa1-7b06-4d20-bb4d-b97a5ee23ce3" event=&{ID:de737447-24b2-4f9e-a805-ad1161a169fb Type:ContainerDied Data:67620
447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7}
Aug 31 05:02:32 ip-10-128-105-64 kubelet[4540]: I0831 05:02:32.453238    4540 scope.go:110] "RemoveContainer" containerID="67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7"
Aug 31 05:02:32 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:32.454276132Z" level=info msg="RemoveContainer for \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\""
Aug 31 05:02:32 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:32.461169671Z" level=info msg="RemoveContainer for \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\" returns successfully"
Aug 31 05:02:32 ip-10-128-105-64 kubelet[4540]: I0831 05:02:32.461324    4540 scope.go:110] "RemoveContainer" containerID="67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7"
Aug 31 05:02:32 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:32.461466477Z" level=error msg="ContainerStatus for \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\" failed" error="rpc error: code = NotFound desc = an error occurr
ed when try to find container \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\": not found"
Aug 31 05:02:32 ip-10-128-105-64 kubelet[4540]: E0831 05:02:32.464644    4540 remote_runtime.go:334] "ContainerStatus from runtime service failed" err="rpc error: code = NotFound desc = an error occurred when try to find container \"67620447e632a1b3edb19fe05a
2b44211cffb996bf6ba5ff52811a8bf32eb8b7\": not found" containerID="67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7"
Aug 31 05:02:32 ip-10-128-105-64 kubelet[4540]: I0831 05:02:32.464681    4540 pod_container_deletor.go:52] "DeleteContainer returned error" containerID={Type:containerd ID:67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7} err="failed to get co
ntainer status \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\": rpc error: code = NotFound desc = an error occurred when try to find container \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\": not found"
Aug 31 05:02:33 ip-10-128-105-64 containerd[4512]: time="2022-08-31T05:02:33.133975568Z" level=error msg="Failed to get usage for snapshot \"67620447e632a1b3edb19fe05a2b44211cffb996bf6ba5ff52811a8bf32eb8b7\"" error="lstat /var/lib/containerd/io.containerd.sna
pshotter.v1.overlayfs/snapshots/302/fs/ide: no such file or directory"
```

After this, the following logs keep appearing over and over again:

```console
Aug 31 05:03:23 ip-10-128-105-64 kubelet[4540]: I0831 05:03:23.117919    4540 pod_container_manager_linux.go:194] "Failed to delete cgroup paths" cgroupName=[kubepods burstable podde737447-24b2-4f9e-a805-ad1161a169fb] err="unable to destroy cgroup paths for c
group [kubepods burstable podde737447-24b2-4f9e-a805-ad1161a169fb] : remove /sys/fs/cgroup/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podde737447_24b2_4f9e_a805_ad1161a169fb.slice: device or resource busy"
```

### What did you expect to happen?

Not use 100% cpu

### How can we reproduce it (as minimally and precisely as possible)?

Not sure

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.9", GitCommit:"b631974d68ac5045e076c86a5c66fba6f128dc72", GitTreeState:"clean", BuildDate:"2022-01-19T17:51:12Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.11-eks-18ef993", GitCommit:"b9628d6d3867ffd84c704af0befd31c7451cdc37", GitTreeState:"clean", BuildDate:"2022-07-06T18:06:23Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}

$ containerd --version
containerd containerd.io 1.6.6 10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1

$ runc --version
runc version 1.1.2
commit: v1.1.2-0-ga916309
spec: 1.0.2-dev
go: go1.17.11
libseccomp: 2.5.1

$ kubelet --version
Kubernetes v1.22.12-eks-ba74326

$ cat /etc/kubernetes/kubelet/kubelet-config.json
{
  "kind": "KubeletConfiguration",
  "apiVersion": "kubelet.config.k8s.io/v1beta1",
  "address": "0.0.0.0",
  "authentication": {
    "anonymous": {
      "enabled": false
    },
    "webhook": {
      "cacheTTL": "2m0s",
      "enabled": true
    },
    "x509": {
      "clientCAFile": "/etc/kubernetes/pki/ca.crt"
    }
  },
  "authorization": {
    "mode": "Webhook",
    "webhook": {
      "cacheAuthorizedTTL": "5m0s",
      "cacheUnauthorizedTTL": "30s"
    }
  },
  "clusterDomain": "cluster.local",
  "hairpinMode": "hairpin-veth",
  "readOnlyPort": 0,
  "cgroupDriver": "systemd",
  "cgroupRoot": "/",
  "featureGates": {
    "RotateKubeletServerCertificate": true
  },
  "protectKernelDefaults": true,
  "serializeImagePulls": false,
  "serverTLSBootstrap": true,
  "tlsCipherSuites": [
    "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256",
    "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305",
    "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305",
    "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384",
    "TLS_RSA_WITH_AES_256_GCM_SHA384",
    "TLS_RSA_WITH_AES_128_GCM_SHA256"
  ],
  "resolvConf": "/run/systemd/resolve/resolv.conf",
  "registryBurst": 20,
  "registryPullQPS": 15,
  "maxPods": 234,
  "imageGCHighThresholdPercent": 70,
  "imageGCLowThresholdPercent": 50,
  "clusterDNS": [
    "172.20.0.10"
  ],
  "evictionHard": {
    "memory.available": "100Mi",
    "nodefs.available": "10%",
    "nodefs.inodesFree": "5%"
  },
  "kubeReserved": {
    "cpu": "110m",
    "ephemeral-storage": "1Gi",
    "memory": "2829Mi"
  }
}
```

</details>

### Cloud provider

<details>
AWS
</details>

### OS version

<details>

```console
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
$ uname -a
Linux ip-10-128-105-64 5.15.0-1017-aws #21~20.04.1-Ubuntu SMP Fri Aug 5 11:44:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by bobbypage** [comment]
Hi there, can you please provide repro steps, we are not seeing this issue in any of the cgroupv2 CI jobs, e.g. https://testgrid.k8s.io/sig-node-containerd#cos-cgroupv2-containerd-e2e

Additionally, the kubelet version here is quite old (1.22). Does this repro on e.g. 1.24 or 1.25?

**Comment by DavidHaltinner** [comment]
This seems that it may be related to other issues being tracked in k8s here https://github.com/kubernetes/kubernetes/issues/112124 as well as OpenShift https://github.com/okd-project/okd/issues/1310
The same messages appear, same CPU usage, as well as some other messages as referenced in these two issues (kubelet_getters.go:300] "Path does not exist" xxxxxxx).

I have 3 clusters on 1.24.0 with the same issue myself, OpenShift 4.11. For me, the control planes run out of memory every 3 days or so. the workers more sporadic. but it seems to relate to the number of pods created, then later deleted, so things such as jobs make it easiest to replicate.

**Comment by bobbypage** [comment]
Hmm, have you seen https://github.com/kubernetes/kubernetes/issues/112124#issuecomment-1245903967, it seems related (systemd is overloaded and unable to stop the unit).

Also, is this issue cgroup v2 related? The OS mentioned above (Ubuntu 20.04.5 does enable cgroupv2 by default).

**Comment by MatthewFagan** [comment]
Yes, we are running cgroups v2. We enabled it in the kernel. We are running on Amazon EKS. The highest k8s version they support is 1.23, and that was only released 4 weeks ago. That is why we are on 1.22. It is not possible to test this config on 1.24 or 1.25. I think it could be an issue with runc and cgroups v2, but am not sure. I do not see how systemd overload could be the problem. We are not running massive loads. In fact under testing we have a very minimal load on the system as a whole.

**Comment by dims** [comment]
@MatthewFagan i would be good to escalate this to the EKS team then?

**Comment by danielcb** [comment]
Hi, one of the above mentioned issues was opened by me. We still have this problem on 4 production clusters. After a few days running a lot of cronjobs kubelet starts consuming multiple 10GBs of RAM and sooner or later gets OOM killed. The problem started to appear after upgrading to v1.24.X on each cluster.

**Comment by bobbypage** [comment]
Can you folks provide a bit more details around the setup / possible repro steps? I tried to repro on a 1.24 Ubuntu 20.04 cgroupv2 cluster but was not able to.

My main questions would be:

* kubelet version
* os version
* Is cgroupv2 enabled or not in the OS image?
* container runtime version (containerd/cri-o) / runc verison
* cgroup driver used
* Repro steps (what the cronjobs are doing)

**Comment by danielcb** [comment]
- Ubuntu 20.4.5 LTS (5.15.0-48-generic)
- kubelet v1.24.4
- cgroupv2 is enabled:
```
$ grep cgroup /proc/filesystems
nodev	cgroup
nodev	cgroup2
```
- CRI-O version 1.24.2
- currently crun version 1.6 (commit: 18cf2efbb8feb2b2f20e316520e0fd0b6c41ef4d), but also happened with all current versions of runc
- cgroup driver systemd

Regarding repro I'm not really sure. As far as I can see it happens on all nodes, but it happens faster as more pods are started. The cronjobs are mostly alpine containers running small bash scripts.

**Comment by dispensable** [comment]
Same issue here with `debian11 buster` and `cgroupv2` enabled.

- kubelet: v1.22.4
- runc: v1.1.4

**Comment by bobbypage** [comment]
@dispensable can you provide the containerd version and cgroup driver used? Also, if you can provide the examples of pods being run used it will help to build a repro.

**Comment by dispensable** [comment]
- containerd: 1.6.8
- cgroup driver: systemd

We have lots of pods running log compression tasks and some jenkins pipeline tasks. They will quit after finished their jobs. Kubelet just cannot clean those finished pod with errorlog:
```
2022-10-10 09:34:23.644137500 I1010 09:34:23.644111    4254 pod_container_manager_linux.go:194] "Failed to delete cgroup paths" cgroupName=[kubepods burstable podcf8cf443-38ef-44fe-9fcf-c5f30d8934b4] err="unable to destroy cgroup paths for cgroup [kubepods burstable podcf8cf443-38ef-44fe-9fcf-c5f30d8934b4] : remove /sys/fs/cgroup/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podcf8cf443_38ef_44fe_9fcf_c5f30d8934b4.slice: device or resource busy"
... and lots of
kubelet_getters.go:300] "Path does not exist" path="/var/lib/kubelet/pods/86cba354-348c-4826-9837-df9c616a8862/volumes"
```
![image](https://user-images.githubusercontent.com/3339663/195246854-3f95e9b5-ca2f-4169-b256-7b25ecb89d7c.png)

Then kubelet going to a high cpu usage. All the pods on this node stuck on Terminating status. Only a reboot or `--force` delete can make this node works for 10 minutes then stuck Terminating again.

Use debian10.2 with  backport kernel 5.10 and enable systemd driver and cgoupv2 works well. We have to reinstall all the nodes to debian10.2 to make our cluster stable.

**Comment by bobbypage** [comment]
> Use debian10.2 with backport kernel 5.10 and enable systemd driver and cgoupv2 works well. We have to reinstall all the nodes to debian10.2 to make our cluster stable.

Hmm, this is weird... so are you saying this is a problem only on Debian 11? Not Debian 10.2 with backported 5.10 kernel + cgroupv2 + same k8s version?

What other Debian 11 / Debian 10 changes are significant here? Perhaps systemd version?

**Comment by dispensable** [comment]
> so are you saying this is a problem only on Debian 11? Not Debian 10.2 with backported 5.10 kernel + cgroupv2 + same k8s version?

Yes, only on Debian11. 

Same kernel/k8s/containerd version with cgroupv2 and systemd driver.
except
In Debian11 we use **crun(1.6)** as runtime (cause cilium can't work with runc -- another wired issue)
In Debian10 we use runc(v.1.1.4) as runtime (sorry for the previous misdirection)

> Is the kernel version different between Debian 11 (where you see this issue) and Debian 10 (with the backported 5.10 kernel)? Trying to understand what Debian 11 / Debian 10 changes are significant here? Perhaps systemd version?

Same kernel version (debian10 use official backport source, 11 use official source default kernel).

Debian10 systemd: 241
Debian11 systemd: 247

**Comment by bobbypage** [comment]
> In Debian11 we use crun(1.6) as runtime (cause cilium can't work with runc -- another wired issue)
In Debian10 we use runc(v.1.1.4) as runtime (sorry for the previous misdirection)

Got it, so the big difference in your case seems to be crun vs runc. Has anyone been able to reproduce with runc + containerd? 

/cc @haircommander @mrunalp since we have several reports of the issue with crun / CRI-O. 

I attempted to repro by bringing up a v1.24.6 on Ubuntu 20.04 with cgroupv2 (kernel 5.15.0-1018-gcp), containerd v1.6.8, runc 1.1.4, systemd cgroup driver and running the following job on a single node cluster:

```
apiVersion: batch/v1
kind: Job
metadata:
  name: busybox
spec:
  template:
    spec:
      containers:
        - name: busybox
          image: busybox
          command: ["sh", "-c", "echo hello1; echo hello2; echo hello3; sleep $((10 + RANDOM % 11));"]
      restartPolicy: Never
  backoffLimit: 4
  completions: 50000
  parallelism: 50
```

I have **not** be able to repro so far...

Has anyone be able to repro on cgroupv2 with containerd/runc and can provide some more detailed repro steps? Would help alot to track down the issue.

**Comment by haircommander** [comment]
cc @giuseppe

**Comment by giuseppe** [comment]
it could be related to crun since it creates a sub-cgroup for the container (differently than what runc does) but I've performed the same test as @bobbypage using CRI-O+crun and I cannot reproduce the issue.

It can be related to the systemd version as well since the cgroup is deleted through systemd.

> Aug 31 05:03:23 ip-10-128-105-64 kubelet[4540]: I0831 05:03:23.117919    4540 pod_container_manager_linux.go:194] "Failed to delete cgroup paths" cgroupName=[kubepods burstable podde737447-24b2-4f9e-a805-ad1161a169fb] err="unable to destroy cgroup paths for c
group [kubepods burstable podde737447-24b2-4f9e-a805-ad1161a169fb] : remove /sys/fs/cgroup/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-podde737447_24b2_4f9e_a805_ad1161a169fb.slice: device or resource busy"

Could you try cleaning up the cgroup manually?  I think something like `systemctl stop kubepods-burstable-podde737447_24b2_4f9e_a805_ad1161a169fb.slice` should work.  Does it work?

**Comment by dispensable** [comment]
Hi, We found our problem was caused by a high disk io. On the busy work node, the containerd mounted disk io was so high that short-life pods' rootfs couldn't be umounted immediately. Those umount reqs was put to the cri clean queen of the kubelet. The airflow/jenkins got a timeout and recreated more tasks to retry made the situation even worse. 

After changed disk type to faster ssd, all problems has gone. Furthermore, we find our s3 based distribution file system was using a cache policy which is not matched with this cluster's workload. Filesystem's cache dir was sharing same device with containerd. The wrong cache policy used a lot of disk io bandwidth which also slowed down the containerd's clean process. We changed filesystem's cache policy. Now even use a hdd disk type with debian11, the high cpu usage of the kubelet is unable to observe.

I don't know if this is also applicable to other scenarios, FYI
